core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableProject.java

```diff
@@ -86,4 +99,65 @@ public Result implement(EnumerableRelImplementor implementor, Prefer pref) {
     // EnumerableCalcRel is always better
     throw new UnsupportedOperationException();
   }
+
+  @Override public Pair<RelTraitSet, List<RelTraitSet>> passThroughTraits(
+      RelTraitSet required) {
+    RelCollation collation = required.getCollation();
+    if (collation == null) {
+      return null;
+    }
+
+    final Mappings.TargetMapping map =
+        RelOptUtil.permutationIgnoreCast(
+            getProjects(), getInput().getRowType());
+    if (checkIfSortIsDefinedOnNonTrivalExpr(map, collation)) {
+      return null;
+    }
+
+    final RelCollation newCollation = collation.apply(map);
+    return Pair.of(required, ImmutableList.of(required.replace(newCollation)));
+  }
+
+  @Override public Pair<RelTraitSet, List<RelTraitSet>> deriveTraits(
+      final RelTraitSet childTraits, final int childId) {
+    RelCollation collation = childTraits.getCollation();
+    if (collation == null) {
+      return null;
+    }
+
+    final Mappings.TargetMapping map =
+        RelOptUtil.permutationIgnoreCast(
+            getProjects(), getInput().getRowType());
+    if (checkIfSortIsDefinedOnNonTrivalExpr(map, collation)) {
+      return null;
```

Comment:
```sql
-- no sort for order by
select a, b*2 as d, c from (select distinct a, b, c from foo) order by a;
```

```sql
-- b is int, no sort for order by
select a, cast(b as bigint) as d, c from (select distinct a, b, c from foo) order by a, d;
```

```sql
-- b is int, need sort for order by
select a, cast(b as varchar) as d, c from (select distinct a, b, c from foo) order by a, d;
```

Reply:
For the last 2 negative test cases, I think you already have similar ones in your test.

Reply:
lot of thanks for your explanation. Now I understand it more:

Trait derivation can pass collations up when such collations are defined on trivial exprs.

Reply:
Now I change the implementation to:

1. traits pass through: return null when at least one collation is defined on non-trivial expr.
2. traits derivation: only return null when all collations are defined on non-trivial exprs. Otherwise derive those collations that are defined on trivial expr.

Reply:
For 2, say the input is sorted by a,b,c.
There is a project on top of that, select a, b*-1 as d, c.
With your approach, it will derive collation on [a, c], which is wrong. Its derived collation should be just [a]. Any non-trivial expr will break the collation.

Reply:
I see. So basically the correctness of ordering is just partial sequence ordering that counts from beginning.

Say sort on [a, b, c, d], then it only guarantees [a], [a, b], [a, b, c] and [a, b, c, d].

And the rational is any non-trivial expr will break the ordering guarantees, thus produce a wrong plan.

Reply:
Will make a change (which is easy), just add a break after encountering first non-trivial expr.

Reply:
Now it generates a correct plan (sort enforced for top projection).